core/src/main/java/org/infinispan/persistence/async/AsyncCacheWriter.java

```diff
@@ -355,16 +432,49 @@ public void run() {
       }
 
       private void retryWork(int maxRetries) {
+         // Even with !failSilently we only try maxRetries times as it's possible that the failure is not due to store availability and this
+         // prevents us repeating the failed operation indefinitely
          for (int attempt = 0; attempt < maxRetries; attempt++) {
             if (attempt > 0 && log.isDebugEnabled())
                log.debugf("Retrying due to previous failure. %s attempts left.", maxRetries - attempt);
 
             try {
+               if (!failSilently) {
+                  availabilityLock.lock();
+                  try {
+                     // It's necessary to check the delegate's availability here as it's possible that it changed after
+                     // the AsyncStoreProcessor was created
+                     if (!delegateAvailable) {
+                        if (stopped) {
+                           log.debugf("Failed to write async modifications to %s as the store is unavailable and stop() was called", actual);
+                           return;
+                        }
+
+                        availability.await();
+                     }
+                  } catch (InterruptedException e) {
+                     log.debugf("%s interrupted: %s", this, e);
+                     Thread.currentThread().interrupt();
+                     break;
+                  } finally {
+                     availabilityLock.unlock();
+                  }
+               }
                AsyncCacheWriter.this.applyModificationsSync(modifications);
                return;
             } catch (Exception e) {
                if (log.isDebugEnabled())
                   log.debug("Failed to process async modifications", e);
+
+               if (!failSilently) {
+                  try {
+                     // Wait for availabilityInterval time to ensure that before the next attempt the delegate's availability
+                     // flag will have been updated and availability.await will be reached
+                     Thread.sleep(configuration.availabilityInterval());
+                  } catch (InterruptedException ie) {
+                     break;
```

Comment:
Missed resetting the interrupt flag.

Reply:
I thought that's what you told me to remove? lol

previously it was:
```
Thread.currentThread().interrupt();
break;
```

Reply:
Ah I must have missed what it was before, sorry for the confusion.